File: DataStructures/Deque/ts/deque.ts
/* 

Deque or Double Ended Queue is a type of queue in which 
insertion and removal of elements can either be performed 
from the front or the rear. Thus, it does not follow 
FIFO rule (First In First Out).

Input Restricted Deque - input is restricted at a single end but allows deletion at both the ends.
Output Restricted Deque - output is restricted at a single end but allows insertion at both the ends.
*/

export class DequeNode<T> {
  value: T;
  next: DequeNode<T> | null;
  prev: DequeNode<T> | null;

  constructor(value: T) {
    this.value = value;
    this.next = null;
    this.prev = null;
  }
}

export class Deque<T> {
  private head: DequeNode<T> | null;
  private tail: DequeNode<T> | null;
  private size: number;

  constructor() {
    this.head = null;
    this.tail = null;
    this.size = 0;
  }

  isEmpty() {
    return this.size === 0;
  }

  addFront(value: T): void {
    const newNode = new DequeNode(value);

    if (this.isEmpty()) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.next = this.head;
      this.head!.prev = newNode;
      this.head = newNode;
    }

    this.size++;
  }

  addRear(value: T): void {
    const newNode = new DequeNode(value);

    if (this.isEmpty()) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.prev = this.tail;
      this.tail!.next = newNode;
      this.tail = newNode;
    }

    this.size++;
  }

  removeFront(): T | null {
    if (this.isEmpty()) {
      return null;
    }

    const removedValue = this.head!.value;
    this.head = this.head!.next;
    if (this.head) {
      this.head.prev = null;
    } else {
      this.tail = null;
    }

    this.size--;
    return removedValue;
  }

  removeRear(): T | null {
    if (this.isEmpty()) {
      return null;
    }

    const removedValue = this.tail!.value;
    this.tail = this.tail!.prev;
    if (this.tail) {
      this.tail.next = null;
    } else {
      this.head = null;
    }

    this.size--;
    return removedValue;
  }

  peekFront(): T | null {
    return this.isEmpty() ? null : this.head!.value;
  }

  peekRear(): T | null {
    return this.isEmpty() ? null : this.tail!.value;
  }

  getSize(): number {
    return this.size;
  }
}

File: DataStructures/Queue/Deque/go/deque.go
package deque

type Deque struct {
	items []int
}

func (d *Deque) isEmpty() bool {
	return len(d.items) == 0
}

func (d *Deque) addFront(item int) {
	d.items = append([]int{item}, d.items...)
}

func (d *Deque) addRear(item int) {
	d.items = append(d.items, item)
}

func (d *Deque) removeFront() int {
	if d.isEmpty() {
		return 0
	}
	front := d.items[0]
	d.items = d.items[1:]
	return front
}

func (d *Deque) removeRear() int {
	if d.isEmpty() {
		return 0
	}

	rear := d.items[len(d.items)-1]
	d.items = d.items[:len(d.items)-1]
	return rear
}

func (d *Deque) size() int {
	return len(d.items)
}

func (d * Deque) peekFront() int {
	if d.isEmpty() {
		return 0
	}
	return d.items[0]
}

func (d * Deque) peekRear() int {
	if d.isEmpty() {
		return 0
	}
	return d.items[len(d.items) - 1]
}